### PR TITLE
Sub pub master

### PR DESCRIPTION
**Pull request by deanlee**
reusing an kj::array to do alignment related copy.
poll() will validate the incoming message,and return an event array
pollRaw() and receive() will not read the event. caller must call getEvent to parse and get the event.
check if message is "alive" when the optional parameter "alive" is set to true.but I have no idea where to use this flag.so I keep the message processing mode remains the same as before.: commaai/openpilot#1476
@pd0wm @adeebshihadeh

### CODE REVIEW

**Comment by deanlee** [comment]
I have tested it and it seems to work fine.

**Comment by pd0wm** [comment]
I'm confused by all the extra inputs and classes needed. Can we make the API exactly the same as the python version? Would be nice we can use the [] operator overloading to get the ` cereal::Event::Reader ` out.

**Comment by deanlee** [comment]
Although I have tried to keep the code as simple as possible， but u r right, we can make the interface clearer and easier to use,I will continue to optimize it.

**Comment by pd0wm** [comment]
I'm fine with loggerd not using submaster since it's use case is so different.

**Comment by deanlee** [comment]
**Examples**
single sub without poller:
```c++
 SubMessage msg(NULL, "sensorEvents");
 cereal::Event::Reader *r = msg.receive();
 if (r) printf("%d", r->getValid());

```
with poller:
 ```c++
SubMaster sm(NULL, {"controlsState", "sensorEvents", "cameraOdometry"});
 sm.update(1000);
 for (cereal::Event::Reader event : sm.allAliveAndValid()){}

```
single pub:
```c++
 PubMessage p(NULL, "sensorEvents");
 p.send(msg);
```
multiple pub:
```c++
 PubMaster p(NULL, {"controlsState", "sensorEvents", "cameraOdometry"});
 p.send("controlsState", msg);
```

**Comment by pd0wm** [comment]
Do we need the extra SubMessage class? I would just put it all in SubMaster. We don't need to add complexity for single socket subscribers.

The all_alive_and_valid method in python returns a bool, not a list of sockets. 

Can you overload the `[]` operator to get messages from the SubMaster?

like
```c++
SubMaster sm( {"controlsState", "sensorEvents", "cameraOdometry"});
sm.update(1000);

if (sm.updated["controlsState"]) {
  cereal::Event::Reader e = sm["controlsState"];
  // do stuff
}
```

**Comment by deanlee** [comment]
updated.all in SubMaster now.
the only difference is using sm.updated("controlsState") instead of sm.update["controlsState"]. 

to support "updated[]",I  need add a new public member variable std::map. do we need that?

**Comment by pd0wm** [comment]
@deanlee Thanks! That looks like the API I had in mind. Updated as a function is fine!

It's a big change, so I'll need some time to review it. Can you make sure your openpilot PR is properly rebased on latest master? Is see some unrelated changes to stuff like `read_db_value` in there.

**Comment by deanlee** [comment]
my pleasure.I have rebased it to latest master branch.

File: messaging/socketmaster.cc [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    SubMessage *m = messages_.at(s);
+    const size_t size = (msg->getSize() / sizeof(capnp::word)) + 1;
+    if (m->buf.size() < size) {
+      m->buf = kj::heapArray<capnp::word>(size);
+    }
+    memcpy(m->buf.begin(), msg->getData(), msg->getSize());
+    delete msg;
+
+    if (m->msg_reader) {
+      m->msg_reader->~FlatArrayMessageReader();
```
Comment (pd0wm) [inline messaging/socketmaster.cc:86]: why not `delete m->msg_reader`?

**Pull request merged**